lib/sensors/MoistureSensor.cpp

```diff
@@ -0,0 +1,93 @@
+#include "mbed.h"
+#include "MoistureSensor.h"
+
+constexpr int Sensor_I2C_Address = 0x36 << 1;                               //MBED I2C uses 8 bit addressing, so addresses are left shifted by 1 (may need to be shifted by 2) 
+
+constexpr int Sensor_Status_Base = 0x00;                                    //Base address registers for different modules
+constexpr int Sensor_Moisture_Base = 0x0F;
+
+constexpr int Sensor_Status_HW_ID = 0x01;                                   //Function address register for the sensor's HW ID
+constexpr int Sensor_HW_ID_Code = 0x68;                                     //Expected value for sensor HW ID; NOTE: This is not the expected value as specified by the datasheet,
+                                                                            //it's just what I found through my own testing. It sometimes fluctuates on its own, so it's hard to trust :(
+
+constexpr int Sensor_Moisture_Function = 0x10;                              //Function address registers for moisture and temperature modules        
+constexpr int Sensor_Temp_Function = 0x04;  
+
+MoistureSensor::MoistureSensor(PinName sda, PinName scl) : i2c_(sda, scl){}
+
+bool MoistureSensor::Is_Initialized(){
+    if(this->Read_HW_ID() == Sensor_HW_ID_Code){                            //compare received HW ID Code to correct one
+        return true;
+    }
+
+    return false;
+}
+
+int MoistureSensor::Read_HW_ID(){
+    char cmd[2];
+    cmd[0] = Sensor_Status_Base;
+    cmd[1] = Sensor_Status_HW_ID;
+
+    char check[1];
+
+    i2c_.write(Sensor_I2C_Address, cmd, 2);                                 //initialize registers for checking device ID
+
+    i2c_.read(Sensor_I2C_Address, check, 1);                                //read device ID
+
+    return check[0];
+}
+
+uint16_t MoistureSensor::Read_Moisture(){
+    /*
+    if(!(this->Is_Initialized())){                                          //checks if device is initialized, returns 65534 if there is an issue
+        return 65534;                                                       //this is commented out because the HW_ID reading isn't consistent currently
+    }
+    */
+
+    char cmd[2];
+    cmd[0] = Sensor_Moisture_Base;
+    cmd[1] = Sensor_Moisture_Function;
+
+    char buf[2];
+    uint16_t ret = 65535;
+
+    do{
+        wait_ms(1);
+
+        i2c_.write(Sensor_I2C_Address, cmd, 2);                             //initialize registers for reading moisture
+
+        wait_ms(1);
+
+        i2c_.read(Sensor_I2C_Address, buf, 2);                              //read moisture
+
+        ret = ((uint16_t)buf[0] << 8 | buf[1]);                             //concatenate bytes together
+
+    } while(ret == 65535);                                                  //repeat until value has been measured
```

Comment:
should add some kind of timeout here to prevent blocking if theres a stuck error

Reply:
As a placeholder, I added a counter to terminate the loop after 10 repetitions so we shouldn't get a stuck error (although it's not a perfect solution either way, and so it can be tweaked).